// pg/lib/defaults.js
'use strict'

let user
try {
  user = process.platform === 'win32' ? process.env.USERNAME : process.env.USER
} catch {
  
}

module.exports = {
  
  host: 'localhost',

  
  user,

  
  database: undefined,

  
  password: null,

  
  
  
  connectionString: undefined,

  
  port: 5432,

  
  
  rows: 0,

  
  binary: false,

  

  
  
  max: 10,

  
  
  idleTimeoutMillis: 30000,

  client_encoding: '',

  ssl: false,

  
  sslnegotiation: undefined,

  application_name: undefined,

  fallback_application_name: undefined,

  options: undefined,

  parseInputDatesAsUTC: false,

  
  
  statement_timeout: false,

  
  
  lock_timeout: false,

  
  
  idle_in_transaction_session_timeout: false,

  
  query_timeout: false,

  connect_timeout: 0,

  keepalives: 1,

  keepalives_idle: 0,
}

const pgTypes = require('pg-types')

const parseBigInteger = pgTypes.getTypeParser(20, 'text')
const parseBigIntegerArray = pgTypes.getTypeParser(1016, 'text')


module.exports.__defineSetter__('parseInt8', function (val) {
  pgTypes.setTypeParser(20, 'text', val ? pgTypes.getTypeParser(23, 'text') : parseBigInteger)
  pgTypes.setTypeParser(1016, 'text', val ? pgTypes.getTypeParser(1007, 'text') : parseBigIntegerArray)
})
